src/stdx/zipfian.zig

```diff
@@ -0,0 +1,490 @@
+//! Zipfian-distributed random number generation.
+//!
+//! In the Zipfian distribution a small percentage of candidate
+//! items have a high probability of being selected, while most items
+//! have very low probability of being selected.
+//! It is commonly understood to model the "80-20" Pareto principle,
+//! and to be a discreet version of the Pareto distribution,
+//! and terminology related to both are often used interchangably.
+//!
+//! Zipfian numbers follow an inverse power law, where the 1st item
+//! is selected with high probability, and subsequent items
+//! quickly fall off in probability. The rate of the fall off
+//! is tunable by the _skew_, also called `s`, or `theta`,
+//! depending on the source.
+//!
+//! Reference:
+//!
+//! - https://en.wikipedia.org/wiki/Zipf's_law#Formal_definition
+//!
+//! Note that it is not actually possible to select a value for
+//! theta that literally follows the "80-20" rule for arbitrary set sizes;
+//! the proportion of items that cumulatively make up 80% probability will
+//! change as the set grows.
+//! A zipfian generator that can adaptively follow the 80-20 rule is left for future work.
+//!
+//! In practice these probabilities often need to be spread across e.g. a
+//! table's keyspace, which involves some kind of mapping step from index to index.
+//! Because that mapping is non-trivial to optimize, it is also provided here.
+//!
+//! The algorithm here is based on
+//! "Quickly Generating Billion-Record Synthetic Databases", Jim Gray et al, SIGMOD 1994.
+//! Per the paper it is adapted from Knuth vol 3.
+//! This is also the algorithm used by YCSB's ZipfianGenerator.java.
+//! Note that the code listing in the paper contains obvious errors,
+//! corrected here and in YCSB.
+//!
+//! There are two generators here,
+//! both of which generate random keys from 0 to a specified maximum.
+//! In the basic `ZipfianGenerator`, key 0 has the highest probability,
+//! 1 the next highest, etc.
+//! The `ShuffledZipfian` generator instead spreads the distribution out
+//! across the key space as if it were a shuffled deck.
+//!
+//! Both generators allow for the key space to grow (but not shrink),
+//! dynamically recomputing the distribution. When the `ShuffledZipfian`
+//! generator grows it acts as if each new key was inserted into the shuffled
+//! deck randomly, preserving the relative probability of existing keys.
+//!
+//! The non-shuffled generator should pass a 2-sample Kolmogorov–Smirnov test;
+//! the shuffled generator does not because the tail distribution is fudged.
+
+const std = @import("std");
+const assert = @import("std").debug.assert;
+const Random = @import("std").Random;
+const math = @import("std").math;
+const BoundedArray = @import("./bounded_array.zig").BoundedArray;
```

Comment:
NIT: It's a common idiom importing the topmost container and derive the other imports from there.

Examples:

Use `const Random = std.Random` instead of `@import("std").Random`.

 Import `const stdx = @import("stdx.zig")` and then use `stdx.BoundedArray` instead of `@import("./bounded_array.zig").BoundedArray`.